src/libYARP_os/src/yarp/os/PeriodicThread.cpp

```diff
@@ -182,29 +243,30 @@ class yarp::os::PeriodicThread::Private : public ThreadImpl
             owner->run();
         }
 
-
         // At the end of each run of updateModule function, the thread is supposed
         // to be suspended and release CPU to other threads.
         // Calling a yield here will help the threads to alternate in the execution.
         // Note: call yield BEFORE computing elapsed time, so that any time spent due to
-        // yield is took into account and the sleep time is correct.
+        // yield is taken into account and the sleep time is correct.
         yield();
 
         lock();
         count++;
-        double elapsed = nowFunc() - currentRun;
+        double now = nowFunc();
+        double elapsed = now - currentRun;
+        double sleepPeriod = delayEstimator->computeDelay(count, now, elapsed);
         //save last
         totalUsed += elapsed;
         sumUsedSq += elapsed * elapsed;
         unlock();
 
-        sleepPeriod = adaptedPeriod - elapsed; // everything is in [seconds] except period, for it is used in the interface as [ms]
-
         delayFunc(sleepPeriod);
     }
 
     void run() override
     {
+        delayEstimator->reset(nowFunc());
```

Comment:
I'm not really sure, but should `reset` be protected by `lock()` and `unlock()` as well?

Reply:
It should not be a problem since `run()` is called only once after thread start. There are other methods that also lack mutex protection such as `getPeriod()`, `resetStat()` and `suspend()/resume()/isSuspended()`. Should I lock everything, just to be 100% sure?